sql/mdl.cc

```diff
@@ -110,37 +110,52 @@ static void init_mdl_psi_keys(void) {
-    {0, "Waiting for locking service lock", 0, PSI_DOCUMENT_ME},
-    {0, "Waiting for spatial reference system lock", 0, PSI_DOCUMENT_ME},
-    {0, "Waiting for acl cache lock", 0, PSI_DOCUMENT_ME},
-    {0, "Waiting for column statistics lock", 0, PSI_DOCUMENT_ME},
-    {0, "Waiting for backup lock", 0, PSI_DOCUMENT_ME},
-    {0, "Waiting for resource groups metadata lock", 0, PSI_DOCUMENT_ME},
-    {0, "Waiting for foreign key metadata lock", 0, PSI_DOCUMENT_ME}};
+const MDL_key::PSI_stage_info_with_name
+    MDL_key::m_namespace_to_wait_state_name[NAMESPACE_END] = {
+        {{0, "Waiting for global read lock", 0, PSI_DOCUMENT_ME},
+         "global read"},
+        {{0, "Waiting for tablespace metadata lock", 0, PSI_DOCUMENT_ME},
+         "tablespace metadata"},
+        {{0, "Waiting for schema metadata lock", 0, PSI_DOCUMENT_ME},
+         "schema metadata"},
+        {{0, "Waiting for table metadata lock", 0, PSI_DOCUMENT_ME},
+         "table metadata"},
+        {{0, "Waiting for stored function metadata lock", 0, PSI_DOCUMENT_ME},
+         "stored function metadata"},
+        {{0, "Waiting for stored procedure metadata lock", 0, PSI_DOCUMENT_ME},
+         "stored procedure metadata"},
+        {{0, "Waiting for trigger metadata lock", 0, PSI_DOCUMENT_ME},
+         "trigger metadata"},
+        {{0, "Waiting for event metadata lock", 0, PSI_DOCUMENT_ME},
+         "event metadata"},
+        {{0, "Waiting for commit lock", 0, PSI_DOCUMENT_ME}, "commit"},
+        {{0, "User lock", 0, PSI_DOCUMENT_ME},
+         /* Be compatible with old status. */ "user"},
+        {{0, "Waiting for locking service lock", 0, PSI_DOCUMENT_ME},
+         "locking service"},
+        {{0, "Waiting for spatial reference system lock", 0, PSI_DOCUMENT_ME},
+         "spatial reference system"},
+        {{0, "Waiting for acl cache lock", 0, PSI_DOCUMENT_ME}, "acl cache"},
+        {{0, "Waiting for column statistics lock", 0, PSI_DOCUMENT_ME},
+         "column statistics"},
+        {{0, "Waiting for backup lock", 0, PSI_DOCUMENT_ME}, "backup"},
+        {{0, "Waiting for resource groups metadata lock", 0, PSI_DOCUMENT_ME},
+         "resource groups metadata"},
+        {{0, "Waiting for foreign key metadata lock", 0, PSI_DOCUMENT_ME},
+         "foreign key metadata"}};
 
 #ifdef HAVE_PSI_INTERFACE
 void MDL_key::init_psi_keys() {
   int i;
   int count;
-  PSI_stage_info *info MY_ATTRIBUTE((unused));
+  const PSI_stage_info *info MY_ATTRIBUTE((unused));
 
   count =
       static_cast<int>(array_elements(MDL_key::m_namespace_to_wait_state_name));
   for (i = 0; i < count; i++) {
     /* mysql_stage_register wants an array of pointers, registering 1 by 1. */
-    info = &MDL_key::m_namespace_to_wait_state_name[i];
-    mysql_stage_register("sql", &info, 1);
+    info = &MDL_key::m_namespace_to_wait_state_name[i].stage_info;
+    mysql_stage_register("sql", const_cast<PSI_stage_info **>(&info), 1);
```

Comment:
Many tests are failing here with:

_build-8.0-ASan/runtime_output_directory/mysqld(__interceptor_backtrace+0x41) [0x9a03c1]
_build-8.0-ASan/runtime_output_directory/mysqld(my_print_stacktrace(unsigned char*, unsigned long)+0xce) [0x2ce5f2e]
_build-8.0-ASan/runtime_output_directory/mysqld(handle_fatal_signal+0x509) [0xeebe59]
libc.so.6(+0x35850) [0x7fb785523850]
_build-8.0-ASan/runtime_output_directory/mysqld(pfs_register_stage_v1(char const*, PSI_stage_info_v1**, int)+0x2a9) [0x3495419]
_build-8.0-ASan/runtime_output_directory/mysqld(MDL_key::init_psi_keys()+0xa7) [0x13cf977]
_build-8.0-ASan/runtime_output_directory/mysqld(mdl_init()+0x14) [0x13cfbe4]
_build-8.0-ASan/runtime_output_directory/mysqld() [0xa37239]
_build-8.0-ASan/runtime_output_directory/mysqld(mysqld_main(int, char**)+0x1617) [0xa32257]
_build-8.0-ASan/runtime_output_directory/mysqld(main+0xd) [0xa1503d]
libc.so.6(__libc_start_main+0x108) [0x7fb78550e858]
_build-8.0-ASan/runtime_output_directory/mysqld() [0x9187db]

Reply:
I looked into this, and the segfault is from not having permission to write into read-only memory.

pfs_register_stage_v1 tries to set info->m_key here:

```
info->m_key = register_stage_class ...
```

Reply:
Sorry, we went too far with declaring `m_namespace_to_wait_state_name` as a `const`.